### PR TITLE
Fix a problem with double initialization

### PR DESCRIPTION
**Pull request by baens**
The check_handle has a problem with double initialization. So instead of trying to initialize on start, do it when the plugin is initialized and just start and stop the listening loops.

This addresses some of #336. There is still problems with the system reconnecting after the 2nd start, but that will probably need to be addressed separately.

### CODE REVIEW

File: src/plugins/node/loop/nodeloopplugin.cpp [hunk 1]
```diff
@@ -136,30 +139,31 @@ extern "C" {
 	}
 
 	NODELOOPPLUGIN_DECL int ibmras_monitoring_plugin_init(const char* properties) {
-		return 0;
+       uv_check_init(uv_default_loop(), &check_handle);
```
Comment (baens) [inline src/plugins/node/loop/nodeloopplugin.cpp]: So, if anyone knows `libuv` better than I do (which honestly I don't know anything about how it works). There is probably a better way of releasing the event loops on start and stop. But this way seems to work ok.

Comment (sam-github) [inline src/plugins/node/loop/nodeloopplugin.cpp]: /to @bnoordhuis?

File: src/plugins/node/loop/nodeloopplugin.cpp [hunk 2]
```diff
@@ -109,7 +112,7 @@ pushsource* createPushSource(uint32 srcid, const char* name) {
 }
 
 void OnCheck(uv_check_t* handle) {
-	const uv_loop_t* const loop = handle->loop;
+    const uv_loop_t* const loop = handle->loop;
```
Comment (bnoordhuis) [inline src/plugins/node/loop/nodeloopplugin.cpp]: Indent error.  You seem to be using spaces whereas everything else in this file uses tabs.

File: src/plugins/node/loop/nodeloopplugin.cpp (hunk 1 above)
Comment (bnoordhuis) [inline src/plugins/node/loop/nodeloopplugin.cpp]: I know a thing or two about libuv but I don't know what "releasing the event loops" means in this context.

File: src/plugins/node/loop/nodeloopplugin.cpp (hunk 2 above)
Comment (baens) [inline src/plugins/node/loop/nodeloopplugin.cpp]: My bad, I will get that fixed.

File: src/plugins/node/loop/nodeloopplugin.cpp (hunk 1 above)
Comment (bnoordhuis) [inline src/plugins/node/loop/nodeloopplugin.cpp]: If I had to guess, I'd say you called `uv_check_init()` again before the `uv_close()` call completed, corrupting libuv's internal structures.  Init once and start/stop on demand is fine though; an inactive check handle consumes zero libuv resources.

File: src/plugins/node/loop/nodeloopplugin.cpp (hunk 2 above)
Comment (baens) [inline src/plugins/node/loop/nodeloopplugin.cpp]: This should now be fixed.

**Comment by baens** [comment]
This should now be ready for merging. The approach I took looks to be an OK approach by way of the conversation had with @bnoordhuis 

If there is anything else that is needed, let me know.

File: src/plugins/node/loop/nodeloopplugin.cpp [hunk 3]
```diff
@@ -68,13 +68,16 @@ static void GetLoopInformation(uv_timer_s *data) {
 #else
 static void GetLoopInformation(uv_timer_s *data, int status) {
 #endif
+    double mean = std::numeric_limits<double>::infinity();
+    if(num != 0)
+        mean = sum / num;
```
Comment (bnoordhuis) [inline src/plugins/node/loop/nodeloopplugin.cpp]: This change seems unrelated?

Comment (baens) [inline src/plugins/node/loop/nodeloopplugin.cpp]: I made that change because when starting and stopping the plugin, there may be no accumulated data. So `num` might be 0 at the time this event is fired, which then causes a divide by 0. This is to protect against that corner case. 

I made that change because when starting and stopping the plugin there might be a time when the accumulated data so `num` would be 0. This is to protect against that.

Comment (bnoordhuis) [inline src/plugins/node/loop/nodeloopplugin.cpp]: Okay.  Can you split it off into a separate commit?  Rule of thumb: one change per commit and each with a commit log explaining what changed and why.

The bulk whitespace update you did makes the diff noisier than it needs to be but it's up to you to decide whether you want to leave it in or not.

Aside: I don't know if the style guide says yes or no on spaces before parens.

**Comment by baens** [comment]
Completely agreed with the comment from @bnoordhuis . So I've now split those two changes into two separated isolated commits. I've hopefully removed any noise due to formatting changes.

**Comment by sjanuary** [comment]
@baens we need a CLA (contributor license agreement) from you before we can accept this pull request. The current process is outlined here: https://github.com/RuntimeTools/appmetrics/blob/master/CONTRIBUTING.md.

**Comment by sjanuary** [comment]
@baens We have simplified the CLA process for Node Application Metrics and now you can just confirm your acceptance by typing a statement into this pull request, e.g. "I confirm my acceptance of the project CLA for Node Application Metrics". CLA is here for reference: https://github.com/RuntimeTools/appmetrics/raw/master/NodeApplicationMetrics-V1-CLA.odt. If you can do that then we will be able to merge this in.

**Comment by baens** [comment]
I confirm my acceptance of the project CLA for Node Application Metrics

~Erik Lindblom.

File: src/plugins/node/loop/nodeloopplugin.cpp [hunk 4]
```diff
@@ -68,13 +68,17 @@ static void GetLoopInformation(uv_timer_s *data) {
 #else
 static void GetLoopInformation(uv_timer_s *data, int status) {
 #endif
+	double mean = std::numeric_limits<double>::infinity();
```
Comment (sjanuary) [inline src/plugins/node/loop/nodeloopplugin.cpp]: I think this should be 0 rather than infinity.  I know technically dividing a number by 0 equals infinity but in terms of calculating mean values using infinity is not right.  It will also look really weird if we try to put this data in a graph.

File: src/plugins/node/loop/nodeloopplugin.cpp [hunk 5]
```diff
@@ -140,30 +140,31 @@ extern "C" {
...
 
 	NODELOOPPLUGIN_DECL int ibmras_monitoring_plugin_init(const char* properties) {
+		uv_check_init(uv_default_loop(), &check_handle);
+		uv_unref(reinterpret_cast<uv_handle_t*>(&check_handle));
+
+		plugin::timer = new uv_timer_t;
+		uv_timer_init(uv_default_loop(), plugin::timer);
+		uv_unref((uv_handle_t*) plugin::timer); // don't prevent event loop exit
+
 		return 0;
```
Comment (sjanuary) [inline src/plugins/node/loop/nodeloopplugin.cpp]: @sam-github I noticed you reviewed this PR, does this look ok to you?  I think we intentionally designed appmetrics not to do any work on init so that there's no overhead if it's not used, I'm wondering if this breaks that requirement?

Comment (sam-github) [inline src/plugins/node/loop/nodeloopplugin.cpp]: It unrefs the handles, and doesn't start them, so I don't think it should have any negative effects.

Comment (sjanuary) [inline src/plugins/node/loop/nodeloopplugin.cpp]: Ok, thanks. I'm going to accept this PR in that case.

**Comment by sjanuary** [comment]
I'm going to accept this PR and then change the default mean of infinity to 0 afterwards as in https://github.com/RuntimeTools/appmetrics/pull/348#discussion_r97959058.
